### PR TITLE
Rewrite ChainerX math tests

### PR DESCRIPTION
**Pull request by asi1024**
Part of #6401.

### CODE REVIEW

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 1]
```diff
@@ -586,33 +586,37 @@ def _create_dummy_array_for_dot(xp, shape, dtype):
 ]
 
 
-@chainerx.testing.numpy_chainerx_array_equal()
-@pytest.mark.parametrize_device(['native:0', 'cuda:0'])
-@pytest.mark.parametrize('input', [
-    numpy.asarray(0), numpy.asarray(-4), numpy.asarray(4),
-    numpy.asarray(-float('inf')), numpy.asarray(float('inf')),
-    numpy.asarray(float('nan')),
-    numpy.full((), 2), numpy.full((0,), 2), numpy.full((2, 3), 2)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Why not leave the previous test patterns?
I think it's far better than random values.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 2]
```diff
@@ -944,33 +950,54 @@ def _create_dummy_array_for_dot(xp, shape, dtype):
...
-    a = dtype_utils.cast_if_numpy_array(xp, a, out_dtype)
-    return xp.exp(a)
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    # Special shapes
+    chainer.testing.product({
+        'shape': [(), (0,), (1,), (2, 0, 3), (1, 1, 1), (2, 3)],
+        'expected_dtypes': _expected_dtypes_math_functions,
+        'input': [0, 2, -2],
+        'check_numpy_strides_compliance': [False],
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Would you mind leaving a comment (either here or in code) why this check must be skipped?

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 3]
```diff
@@ -944,33 +950,54 @@ def _create_dummy_array_for_dot(xp, shape, dtype):
...
+        'input': [float('inf'), -float('inf'), float('nan')],
+        'skip_backward_test': [True],
+        'skip_double_backward_test': [True],
+    })
+))
+class TestExp(UnaryMathTestBase, op_utils.NumpyOpTest):
 
+    def func(self, xp, a):
+        out = xp.exp(a)
+        return dtype_utils.cast_if_numpy_array(xp, out, self.out_dtype)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Is it possible to call `cast_if_numpy_array` in the `UnaryMathTestBase`?

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 4]
```diff
@@ -1024,19 +1058,28 @@ def test_logsumexp_invalid(device, a_shape, axis, keepdims, dtype):
...
-        a = a.astype(out_dtype)
-        # Default is the second axis
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize_pytest('shape,axis', _logsumexp_params)
+@chainer.testing.parameterize_pytest(
+    'dtype,out_dtype', _expected_dtypes_math_functions)
+class TestLogSoftmax(UnaryMathTestBase, op_utils.NumpyOpTest):
+
+    input = 'random'
+    check_numpy_strides_compliance = False
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Ditto regarding comment.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py (hunk 2 above)
Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: `dtype_utils.cast_if_numpy_array` sometimes changes the shape of given numpy arrays if the size is equal to 0.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py (hunk 3 above)
Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Fixed.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py (hunk 4 above)
Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Removed this line.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py (hunk 2 above)
Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: In that case, could you condition the flag based on the size dynamically (or create a slightly more complicated parameterization...)?

Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Ah, you fixed it. Thanks!

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
@niboshi do you have comments on this PR?

**Comment by hvy** [comment]
CI failures seem unrelated. LGTM!

**Pull request merged**